### PR TITLE
[MM-47800] Dismiss profile popover when opening a modal through keyboard shortcut

### PR DESCRIPTION
**Pull request by kostaspt**
#### Summary
Look if a new modal is opened and then hide the profile popover so they don't overlap with each other

#### Ticket Link
https://mattermost.atlassian.net/browse/MM-47800

#### Related Pull Requests
N/A

##### Before

https://user-images.githubusercontent.com/684680/208700498-e2b53d96-76a2-4ad6-b997-2e60e08218cc.mp4

##### After

https://user-images.githubusercontent.com/684680/208700129-82822370-ad5e-455e-8723-98b7be0a5d54.mp4

#### Release Note
```release-note
NONE
```

[MM-47800]: https://mattermost.atlassian.net/browse/MM-47800?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Review by cyrilzhang-mm — CHANGES_REQUESTED** [review]
Looks good, but can we add it to the groups popover too?

**Review by cyrilzhang-mm — APPROVED** [review]

File: components/profile_popover/profile_popover.tsx [hunk 1]
```diff
@@ -234,6 +236,13 @@ ProfilePopoverState
             },
         ));
     }
+
+    componentDidUpdate(prevProps: ProfilePopoverProps) {
+        if (this.props.openModalCount !== prevProps.openModalCount) {
+            this.props.hide?.();
+        }
```
Comment (mkraft) [inline components/profile_popover/profile_popover.tsx:243]: Does this need to know the modal *count* or does it just need to know that *any* modal is open (using the `isAnyModalOpen` selector)?

File: components/user_group_popover/useShouldClose.ts [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+
+    useEffect(() => {
+        setInitialOpenModalCount(openModalCount);
+    }, []);
+
+    useEffect(() => {
+        if (initialOpenModalCount !== openModalCount) {
+            setShouldClose(true);
+        }
+    }, [initialOpenModalCount, openModalCount]);
```
Comment (mkraft) [inline components/user_group_popover/useShouldClose.ts:22]: I commented out this `useEffect` block and the it doesn't seem to change the behaviour... or did I test it wrong?

File: components/profile_popover/profile_popover.tsx (hunk 1 above)
Comment (kostaspt) [reply components/profile_popover/profile_popover.tsx:243]: I thought of the edge case that we'll need that info on a modal. In this case this way will monitor for change (closing the modal) and dismiss it as well. Maybe we could opt in using the selector if there isn't a possible case at the moment or near future. What do you think?

File: components/user_group_popover/useShouldClose.ts (hunk 2 above)
Comment (kostaspt) [reply components/user_group_popover/useShouldClose.ts:22]: This is for the group info, after @cyrilzhang-mm suggestion. Maybe you've checked the regular profile one? It shouldn't work without it.

Comment (mkraft) [reply components/user_group_popover/useShouldClose.ts:22]: Yes you are right. Thank you.

File: components/profile_popover/profile_popover.tsx (hunk 1 above)
Comment (mkraft) [reply components/profile_popover/profile_popover.tsx:243]: I don't understand.

Comment (kostaspt) [reply components/profile_popover/profile_popover.tsx:243]: This is a popover showing information about a user or a group. This is something that could potentially exist in a popup. In that case, there will be a popup open, and by counting modals it would work in that case. However this could be an edge case.

Comment (mkraft) [reply components/profile_popover/profile_popover.tsx:243]: For some reason I'm completely not understanding.

Comment (kostaspt) [reply components/profile_popover/profile_popover.tsx:243]: Maybe because it's not a valid reason 😅

Let's take the insights page as an example. At some point, we decide that we don't want onClick to navigate to DM with the user but to show user information using the profile popover.

image

In this case, there is one modal, so the `isAnyModalOpen` will return `true`. Pressing `cmd+K` or `cmd+/` will trigger another modal:

https://user-images.githubusercontent.com/684680/209405676-f81faa31-0c0a-4524-a84b-49ac05bb882d.mp4

So in the current implementation, it will go from 1 modal to 2 modals so it will hide the popover. But watching the `isAnyModalOpen` it will remain `true` so the popover will remain visible and have the same issue as right now.

I'm probably overthinking it, and there isn't a current issue with that I could find. It's just seems to me a bit more future-proof to me, but I don't have a strong opinion I could probably reuse the existing solution, which would be pretty solid for 99.9% of the cases

Comment (M-ZubairAhmed) [reply components/profile_popover/profile_popover.tsx:243]: i am with @mkraft on this, and i have gone through the above comments, i do think we are probably over engineering, but isAnyModalOpen can work here :P

Comment (kostaspt) [reply components/profile_popover/profile_popover.tsx:243]: Thanks, both for the suggestion. I was definitely overthinking this situation.

File: components/user_group_popover/user_group_popover.tsx [hunk 3]
```diff
@@ -119,6 +122,12 @@ const UserGroupPopover = (props: Props) => {
         }
     }, [searchTerm, doSearch]);
 
+    useEffect(() => {
```
Comment (M-ZubairAhmed) [inline components/user_group_popover/user_group_popover.tsx:125]: Very neat implementation

**Review by M-ZubairAhmed — APPROVED** [review]
LTGM

**Review by mkraft — APPROVED** [review]

**Review by furqanmlk — APPROVED** [review]
Working fine
